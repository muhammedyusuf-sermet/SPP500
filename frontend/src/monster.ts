export enum Size {
	Tiny = "Tiny",
	Small = "Small",
	Medium = "Medium",
	Large = "Large",
	Huge = "Huge",
	Gargantuan = "Gargantuan"
}

export enum MonsterType {
	Beast = "Beast",
	Dragon = "Dragon",
	Monstrosity = "Monstrosity",
	Humanoid = "Humanoid",
	Fiend = "Fiend",
	Undead = "Undead",
	Giant = "Giant",
	Elemental = "Elemental",
	SwarmOfTinyBeasts = "SwarmOfTinyBeasts",
	Construct = "Construct",
	Celestial = "Celestial",
	Aberration = "Aberration",
	Fey = "Fey",
	Plant = "Plant",
	Ooze = "Ooze"
}

export enum MonsterRace {
	AnyRace = "AnyRace",
	Devil = "Devil",
	Demon = "Demon",
	Human = "Human",
	Shapechanger = "Shapechanger",
	Goblinoid = "Goblinoid",
	Titan = "Titan",
	Gnoll = "Gnoll",
	Gnome = "Gnome",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Orc = "Orc",
	Kobold = "Kobold",
	Lizardfolk = "Lizardfolk",
	Merfolk = "Merfolk",
	Sahuagin = "Sahuagin",
	Grimlock = "Grimlock"
}

export enum Alignment {
	Unaligned = "Unaligned",
	AnyAlignment = "AnyAlignment",
	AnyNonGoodAlignment = "AnyNonGoodAlignment",
	AnyNonEvilAlignment = "AnyNonEvilAlignment",
	AnyNonLawfulAlignment = "AnyNonLawfulAlignment",
	AnyNonChaoticAlignment = "AnyNonChaoticAlignment",
	AnyGoodAlignment = "AnyGoodAlignment",
	AnyEvilAlignment = "AnyEvilAlignment",
	AnyLawfulAlignment = "AnyLawfulAlignment",
	AnyChaoticAlignment = "AnyChaoticAlignment",
	AnyNeutralAlignment = "AnyNeutralAlignment",
	LawfulGood = "LawfulGood",
	LawfulNeutral = "LawfulNeutral",
	LawfulEvil = "LawfulEvil",
	NeutralGood = "NeutralGood",
	Neutral = "Neutral",
	NeutralEvil = "NeutralEvil",
	ChaoticGood = "ChaoticGood",
	ChaoticNeutral = "ChaoticNeutral",
	ChaoticEvil = "ChaoticEvil"
}

export enum Environment {
	Arctic = "Arctic",
	Coastal = "Coastal",
	Desert = "Desert",
	Forest = "Forest",
	Grassland = "Grassland",
	Hill = "Hill",
	Mountain = "Mountain",
	Swamp = "Swamp",
	Underdark = "Underdark",
	Underwater = "Underwater",
	Urban = "Urban"
}

export enum MonsterAction {
	SpecialAbility = "SpecialAbility",
	Action = "Action",
	LegendaryAction = "LegendaryAction"
}

export interface IMonsterAbilities {
	name: string,
	description: string,
}

export enum TypeOfAction {
	Ranged = "Ranged",
	Melee = "Melee",
}

export interface IMonsterActions {
	name: string,
	description: string,
	usageCount: number,
	legendaryAction: boolean,
	typeOfAction?: TypeOfAction,
	toHit: number,
	reach: number,
	targets: number,
	averagedDamage: number,
	damageDice: number,
	damageConstant: number,

}

export interface IMonsterState {
	Id?: number,
	Name: string,
	Size?: Size | string;
	Type?: MonsterType | string;
	Race?: MonsterRace | string;
	Environment?: Environment | string;
	Alignment?: Alignment | string;

	ArmorClass?: number;
	HitPoints?: number;
	HitPointDistribution?: string;

	Speed?: string;
	Languages?: string;

	DamageVulnerabilities?: string;
	DamageResistances?: string;
	DamageImmunities?: string;
	ConditionImmunities?: string;

	ChallengeRating?: number;

	AbilityScores: IMonsterAbilityScoreState;
	Senses: SenseMap;
	Skills: SkillMap;
	SavingThrows: IMonsterSavingThrowState;
	//Actions: IActionState[];
}

export interface SkillMap {
	[skillName: string]: number|undefined;
}

export interface SenseMap {
	[senseName: string]: number|undefined;
}

export interface IMonsterAbilityScoreState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IMonsterSavingThrowState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IActionState {
	Name: string,
	Description: string,
	HitBonus?: number,
	Damage?: string,
	DamageBonus?: number,
	Type?: MonsterAction
}
/*export interface IMonster {
	name: string,
	type: MonsterType,
	alignment: MonsterAlignment,
	size: MonsterSize,
	race: MonsterRace,
	environment: MonsterEnvironment,
	resistance?: string,
	damageImmunity?: string,
	conditionImmunity?: string,
	vulnerability?: string,
	armorClass: number,
	hitPoints?: number,
	hitPointDice?: string,
	hitPointDiceAdd?: number,
	speedLand: number,
	speedSwim?: number,
	strStat: number,
	dexStat: number,
	conStat: number,
	intStat: number,
	wisStat: number,
	chaStat: number,
	strSavingThrow?: number,
	dexSavingThrow?: number,
	conSavingThrow?: number,
	intSavingThrow?: number,
	wisSavingThrow?: number,
	chaSavingThrow?: number,
	skillsAthletics: number,
	skillsAcrobatics?: number,
	skillsSleightOfHand?: number,
	skillsStealth?: number,
	skillsArcana?: number,
	skillsHistory?: number,
	skillsInvestigation?: number,
	skillsNature?: number,
	skillsReligion?: number,
	skillsAnimalHandling?: number,
	skillsInsight?: number,
	skillsMedicine?: number,
	skillsPerception?: number,
	skillsSurvival?: number,
	skillsDeception?: number,
	skillsIntimidation?: number,
	skillsPerformance?: number,
	skillsPersuasion?: number,
	sensesBlindsight?: number,
	sensesDarkvision?: number,
	sensesTremorsense?: number,
	sensesTruesight?: number,
	sensesPassivePerception?: number,
	sensesPassiveInvestigation?: number,
	sensesPassiveInsight?: number,
	languages?: string,
	challengeRating: number,
	experiencePoints: number,
	abilities?: IMonsterAbilities[],
	actions?: IMonsterActions[],
}*/