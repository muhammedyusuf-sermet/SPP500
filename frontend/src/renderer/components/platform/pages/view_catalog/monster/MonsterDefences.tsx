import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { isDeepStrictEqual } from 'util';
import { ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Typography, FormControl, InputLabel, Input, FormHelperText, Grid } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

export interface IMonsterDefencesProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		ArmorClass?: number;
		HitPoints?: number;
		HitPointDistribution?: string;
	}
}

export interface IMonsterDefencesState {
	// data
	ArmorClass?: number;
	HitPoints?: number;
	HitPointDistribution?: string;
	// errors
	ArmorClassError?: string;
	HitPointsError?: string;
	HitPointDistributionError?: string;
}

export class MonsterDefences extends React.Component<IMonsterDefencesProps, IMonsterDefencesState> {
	constructor(props: IMonsterDefencesProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterDefencesProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false) {
			this.setState({
				...nextProps.initial
			});
		} else if (this.props.disabled != nextProps.disabled) {
			this.forceUpdate();
		}
	}

	stringToNumber = (toConvert : string) => {
		return isNaN(parseInt(toConvert)) ? undefined : parseInt(toConvert);
	}

	handleNumberChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = this.stringToNumber(event.target.value);
		const name = event.currentTarget.name;
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	handleStringChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		let value: string|undefined = event.target.value;
		if (/^\s*$/.test(value))
			value = undefined
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [event.currentTarget.name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[event.currentTarget.name]: value,
					[event.currentTarget.name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	render() {
		return (
			<ExpansionPanel defaultExpanded CollapseProps={{timeout: 100}} >
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Typography className='heading' >Armor Class and Hit Points</Typography>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Grid container spacing={8} >
						<Grid item xs={12}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="ArmorClass">Armor Class</InputLabel>
								<Input
									id='ArmorClass'
									type='number'
									value={this.state.ArmorClass != undefined ? this.state.ArmorClass : ''}
									name='ArmorClass'
									onChange={this.handleNumberChange}
									aria-describedby="ArmorClass-helper-text" />
								<FormHelperText error id="ArmorClass-helper-text">{this.state.ArmorClassError}</FormHelperText>
							</FormControl>
						</Grid>
						<Grid item xs={6}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="HitPoints">Hit Points</InputLabel>
								<Input
									id='HitPoints'
									type='number'
									value={this.state.HitPoints != undefined ? this.state.HitPoints : ''}
									name='HitPoints'
									onChange={this.handleNumberChange}
									aria-describedby="HitPoints-helper-text" />
								<FormHelperText error id="HitPoints-helper-text">{this.state.HitPointsError}</FormHelperText>
							</FormControl>
						</Grid>
						<Grid item xs={6}>
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor="HitPointDistribution">Hit Point Distribution</InputLabel>
								<Input
									id='HitPointDistribution'
									type='text'
									value={this.state.HitPointDistribution || ''}
									name='HitPointDistribution'
									onChange={this.handleStringChange}
									aria-describedby="HitPointDistribution-helper-text" />
								<FormHelperText error id="HitPointDistribution-helper-text">{this.state.HitPointDistributionError}</FormHelperText>
							</FormControl>
						</Grid>
					</Grid>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		);
	}
}