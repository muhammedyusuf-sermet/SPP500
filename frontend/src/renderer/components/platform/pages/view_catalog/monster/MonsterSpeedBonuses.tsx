import * as React from 'react';
//TODO use validation to ensure the speeds make sense
//const Joi = require('joi');
//import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';
import { Tile, Subtitle, Field, Label, Control, Input } from 'bloomer';

export interface IMonsterSpeedBonusesProps {
	disabled?: boolean;
	// validation TODO not done yet
	//PayloadSchema: JoiObject;
	//ValidationOptions: ValidationOptions;
	// initial values
	initial: {
		[SpeedName: string]: number | undefined;
	}
}

export interface IMonsterSpeedBonusesState {
	[key: string]: number | string | undefined;
}

export class MonsterSpeedBonuses extends React.Component<IMonsterSpeedBonusesProps, IMonsterSpeedBonusesState> {
	constructor(props: IMonsterSpeedBonusesProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	//TODO: Make speed its own dictionary of all the different monvement options.
	//private keyNames: string[] = ["SpeedLand", "SpeedSwim"];

	stringToNumber = (toConvert : string) => {
		return isNaN(parseInt(toConvert)) ? undefined : parseInt(toConvert);
	}

	handleMonsterSpeedChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = this.stringToNumber(event.target.value);
		const name = event.currentTarget.name;
		this.setState({
			[name]: value
		});
		/*TODO: Add validation and help messages
		Joi.validate(
			{ [name]: value },
			Joi.reach(this.props.PayloadSchema, ['Speeds']),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
			});*/
	}

	render() {
		return (
			<Tile className="box" isVertical>
				<Subtitle>Movement Speed</Subtitle>
				<Field isGrouped='centered' isHorizontal>
					<Control isExpanded>
						<Label>Land Speed</Label>
						<Input
							id='SpeedLand'
							type='number'
							placeholder='Land Speed'
							autoComplete='SpeedLand'
							value={this.state.SpeedLand != undefined ? this.state.SpeedLand : ''}
							name='SpeedLand'
							onChange={this.handleMonsterSpeedChange} />
					</Control>
					<Control isExpanded>
						<Label>Swimming Speed</Label>
						<Input
							id='SpeedSwim'
							type='number'
							placeholder='Swimming Speed'
							autoComplete='SpeedSwim'
							value={this.state.SpeedSwim != undefined ? this.state.SpeedSwim : ''}
							name='SpeedSwim'
							onChange={this.handleMonsterSpeedChange} />
					</Control>
				</Field>
			</Tile>
		);
	}
}