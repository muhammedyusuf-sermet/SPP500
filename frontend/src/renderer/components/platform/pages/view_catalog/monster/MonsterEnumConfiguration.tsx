import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { MonsterType, MonsterRace, Size, Environment, Alignment } from '../../../../../../monster';
import { isDeepStrictEqual } from 'util';
import { ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Typography, FormControl, InputLabel, Input, FormHelperText, Grid, Select, Tooltip } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import {Help} from "bloomer";

const types = Object.values(MonsterType);
const sizes = Object.values(Size);
const races = Object.values(MonsterRace);
const environments = Object.values(Environment);
const alignments = Object.values(Alignment);

export interface IMonsterDropdownProps {
	disabled?: boolean,
	name: string,
	selected?: string,
	options: string[]
	onChange: (name: string, selectOption?: string) => void
}

export interface IMonsterDropdownState {
	selected: string
}

export class MonsterDropdown extends React.Component<IMonsterDropdownProps, IMonsterDropdownState> {
	constructor(props: IMonsterDropdownProps) {
		super(props);
		this.state = {
			selected: props.selected ? props.selected : 'Default'
		};
	}

	componentWillReceiveProps(nextProps: IMonsterDropdownProps) {
		if (this.state.selected != nextProps.selected)
			this.setState({
				selected: nextProps.selected ? nextProps.selected : 'Default'
			});
	}

	handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
		this.setState({selected: event.target.value});
		this.props.onChange(this.props.name, event.target.value);
	}

	render() {
		return (
			<Select
				native={true}
				id={this.props.name}
				disabled={this.props.disabled}
				onChange={this.handleChange}
				value={this.state.selected}
				input={<Input name={this.props.name} id={this.props.name} />} >
				<option
					id={this.props.name+'.'+'Default'}
					value='' >
					Default
				</option>
				{this.props.options.map(option =>
					<option
						key={this.props.name+'.'+option}
						id={this.props.name+'.'+option}
						value={option}>
						{option.replace(/([A-Z])/g, ' $1').trim()}
					</option>
				)}
			</Select>
		);
	}
}

export interface IMonsterEnumConfigurationProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		Size?: string;
		Type?: string;
		Race?: string;
		Alignment?: string;
		Environment?: string;
	}
}

export interface IMonsterEnumConfigurationState {
	// data
	Size?: string;
	Type?: string;
	Race?: string;
	Alignment?: string;
	Environment?: string;
	// errors
	SizeError?: string;
	TypeError?: string;
	RaceError?: string;
	AlignmentError?: string;
	EnvironmentError?: string;
}

export class MonsterEnumConfiguration extends React.Component<IMonsterEnumConfigurationProps, IMonsterEnumConfigurationState> {
	constructor(props: IMonsterEnumConfigurationProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterEnumConfigurationProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false)
			this.setState({
				...nextProps.initial
			});
	}

	handleMonsterEnumChange = (name: string, newEnum?: string) => {
		if (newEnum) {
			Joi.validate(
				newEnum,
				Joi.reach(this.props.PayloadSchema, [name]),
				this.props.ValidationOptions,
				(errors: ValidationError) => {
					if(errors) {
						this.setState({
							[name+'Error']: errors.details[0].message
						});
					} else {
						this.setState({
							[name]: newEnum,
							[name+'Error']: undefined
						});
					}
				});
		} else {
			this.setState({
				[name]: undefined,
				[name+'Error']: undefined
			});
		}
	}

	render() {
		return (
			<ExpansionPanel defaultExpanded CollapseProps={{timeout: 100}} >
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Typography className='heading' >Basic Configurations</Typography>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Grid container spacing={8} >
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Type' >Type</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Type}
									name='Type'
									options={types}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Type-helper-text' >{this.state.TypeError}</FormHelperText>
								<Tooltip disableFocusListener title={"Creature types are rough categories of creatures which determine the way game mechanics affect the creature. The choice of type is important, as all creatures which have a given type will share certain characteristics (with some exceptions)."}>
										<Help>Hover for More About Monster Type</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Size' >Size</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Size}
									name='Size'
									options={sizes}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Size-helper-text' >{this.state.SizeError}</FormHelperText>
								<Tooltip disableFocusListener title={"A creature's size determines how much space it occupies on squares or hexes, as shown in the Creature Size and Space table. If the miniature you use for a monster takes up an amount of space different from what's on the table, that's fine, but treat the monster as its official size for all other rules."}>
										<Help>Hover for More About Monster Size</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={4} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Race' >Race</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Race}
									name='Race'
									options={races}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Race-helper-text' >{this.state.RaceError}</FormHelperText>
								<Tooltip disableFocusListener title={"A character race is a fundamental part of the identity and nature of characters in the Dungeons & Dragons role-playing game. Each race has a distinct appearance, behavior and often range of statistics associated with it. See your DM handbook for more information."}>
										<Help>Hover for More About Monster Race</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={6} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Alignment' >Alignment</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Alignment}
									name='Alignment'
									options={alignments}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Alignment-helper-text' >{this.state.AlignmentError}</FormHelperText>
								<Tooltip disableFocusListener title={"Alignment is a categorization of the ethical and moral perspective of player characters, non-player characters, and creatures. Most versions of the game feature a system in which players make two choices for characters. One is the character's views on 'law' vs 'chaos', the other on 'good' vs 'evil'."}>
										<Help>Hover for More About Monster Alignment</Help>
								</Tooltip>
							</FormControl>
						</Grid>
						<Grid item xs={6} >
							<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
								<InputLabel htmlFor='Environment' >Environment</InputLabel>
								<MonsterDropdown
									disabled={this.props.disabled}
									selected={this.state.Environment}
									name='Environment'
									options={environments}
									onChange={this.handleMonsterEnumChange} />
								<FormHelperText error id='Environment-helper-text' >{this.state.EnvironmentError}</FormHelperText>
								<Tooltip disableFocusListener title={"Monster environment is the type of environment that the monster is most likely to be found in."}>
										<Help>Hover for More About Monster Environment</Help>
								</Tooltip>
							</FormControl>
						</Grid>
					</Grid>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		);
	}
}