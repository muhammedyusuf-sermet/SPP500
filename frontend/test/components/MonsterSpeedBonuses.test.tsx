import * as React from 'react';
import { mount, ReactWrapper, shallow } from 'enzyme';

import { IMonsterSpeedBonusesProps, IMonsterSpeedBonusesState, MonsterSpeedBonuses } from '../../src/renderer/components/platform/pages/view_catalog/monster/MonsterSpeedBonuses';
// TODO: test for validation when the speed component actually uses it.
//const Joi = require('joi');
//import { ValidationOptions } from 'joi';

/*const payloadSchema = Joi.object({
	Speed: Joi.object().pattern(
		Joi.symbol().valid(["SpeedLand", "SpeedSwim"]),
		Joi.number().integer().greater(0).allow(0)
	).default({})
});
const validateOptions: ValidationOptions = {
	abortEarly: false,
	convert: true,
	allowUnknown: false,
	context: {

	}
};*/

////// Happy Path //////
describe('Monster SpeedBonuses', () => {

	let monsterSpeedBonusesInstance: ReactWrapper<IMonsterSpeedBonusesProps, IMonsterSpeedBonusesState, MonsterSpeedBonuses>;

	describe('Provided Stat values', () => {

		beforeEach(() => {
			monsterSpeedBonusesInstance =
				mount<MonsterSpeedBonuses, IMonsterSpeedBonusesProps, IMonsterSpeedBonusesState>(
					<MonsterSpeedBonuses
						disabled={false}
						//PayloadSchema={payloadSchema}
						//ValidationOptions={validateOptions}
						initial={{
							"SpeedLand": 20,
							"SpeedSwim": 40,
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterSpeedBonusesInstance).toBeDefined();
		});

		it('should have the correct number for SpeedLand', () => {
			expect(monsterSpeedBonusesInstance.find('input#SpeedLand').props().value).toEqual(20);
		});

		it('should have the correct number for SpeedSwim', () => {
			expect(monsterSpeedBonusesInstance.find('input#SpeedSwim').props().value).toEqual(40);
		});
	});

	describe('componentWillReceiveProps()', () => {
		beforeEach(() => {
			monsterSpeedBonusesInstance =
				mount<MonsterSpeedBonuses, IMonsterSpeedBonusesProps, IMonsterSpeedBonusesState>(
					<MonsterSpeedBonuses
						disabled={false}
						//PayloadSchema={payloadSchema}
						//ValidationOptions={validateOptions}
						initial={{
							"SpeedLand": 20,
							"SpeedSwim": 40,
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterSpeedBonusesInstance).toBeDefined();
		});

		it('renders new props', () => {
			monsterSpeedBonusesInstance.setProps({
				initial: {
					"SpeedLand": 15,
					"SpeedSwim": 30,
				}
			})
			expect(monsterSpeedBonusesInstance.find('input#SpeedLand').props().value).toEqual(15);
			expect(monsterSpeedBonusesInstance.find('input#SpeedSwim').props().value).toEqual(30);
		});

		it('renders without crashing with same props', () => {
			monsterSpeedBonusesInstance.setProps({
				initial: {
					"SpeedLand": 20,
					"SpeedSwim": 40,
				}
			});
			expect(monsterSpeedBonusesInstance.find('input#SpeedLand').props().value).toEqual(20);
			expect(monsterSpeedBonusesInstance.find('input#SpeedSwim').props().value).toEqual(40);
		});
	});

	describe('Small snapshot', () => {
		it ('matches snapshot', () => {
			const shallowMonsterSpeedBonusesInstance =
				shallow<MonsterSpeedBonuses, IMonsterSpeedBonusesProps, IMonsterSpeedBonusesState>(
					<MonsterSpeedBonuses
						disabled={false}
						//PayloadSchema={payloadSchema}
						//ValidationOptions={validateOptions}
						initial={{}}/>
				);
			expect(shallowMonsterSpeedBonusesInstance).toMatchSnapshot();
		});
	});

	describe('Happy Path', () => {

		beforeEach(() => {
			monsterSpeedBonusesInstance =
				mount<MonsterSpeedBonuses, IMonsterSpeedBonusesProps, IMonsterSpeedBonusesState>(
					<MonsterSpeedBonuses
						disabled={false}
						//PayloadSchema={payloadSchema}
						//ValidationOptions={validateOptions}
						initial={{}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterSpeedBonusesInstance).toBeDefined();
		});

		it('should be able to update state', () => {
			monsterSpeedBonusesInstance.find('input#SpeedLand').simulate('change', { target: { value: 30 } })
			monsterSpeedBonusesInstance.find('input#SpeedSwim').simulate('change', { target: { value: 45 } })
			expect(monsterSpeedBonusesInstance.state()).toEqual({
				// data
				SpeedLand: 30,
				SpeedSwim: 45,
				// errors
				SpeedLandError: undefined,
				SpeedSwimError: undefined,
			});
		});
		// STATE CHANGE
		it('should change only SpeedLand when SpeedLand is changed', () => {
			monsterSpeedBonusesInstance.find('input#SpeedLand').simulate('change', { target: { value: '15' } })
			expect(monsterSpeedBonusesInstance.state().SpeedLand).toEqual(15);
		})

		it('should change only SpeedSwim when SpeedSwim is changed', () => {
			monsterSpeedBonusesInstance.find('input#SpeedSwim').simulate('change', { target: { value: '25' } })
			expect(monsterSpeedBonusesInstance.state().SpeedSwim).toEqual(25);
		})

		// DEFAULTS
		it('should change SpeedLand back to undefined when input is empty', () => {
			monsterSpeedBonusesInstance.find('input#SpeedLand').simulate('change', { target: { value: '' } })
			expect(monsterSpeedBonusesInstance.state().SpeedLand).toEqual(undefined);
		})

		it('should change SpeedSwim back to undefined when input is empty', () => {
			monsterSpeedBonusesInstance.find('input#SpeedSwim').simulate('change', { target: { value: '' } })
			expect(monsterSpeedBonusesInstance.state().SpeedSwim).toEqual(undefined);
		})

		// TODO: test validation when its actually implemented
		// VALIDATION
		/*it('should validate SpeedLand when SpeedLand is changed', () => {
			monsterSpeedBonusesInstance.find('input#SpeedLand').simulate('change', { target: { value: '-1'  } })
			expect(monsterSpeedBonusesInstance.state().SpeedLand).toEqual(-1);
			expect(monsterSpeedBonusesInstance.find('Help#SpeedLand').text()).toEqual('\"SpeedLand\" must be greater than 0');
		})

		it('should validate SpeedSwim when SpeedSwim is changed', () => {
			monsterSpeedBonusesInstance.find('input#SpeedSwim').simulate('change', { target: { value: '-9'  } })
			expect(monsterSpeedBonusesInstance.state().SpeedSwim).toEqual(-9);
			expect(monsterSpeedBonusesInstance.find('Help#SpeedSwim').text()).toEqual('\"SpeedSwim\" must be greater than 0');
		})*/
	});
});