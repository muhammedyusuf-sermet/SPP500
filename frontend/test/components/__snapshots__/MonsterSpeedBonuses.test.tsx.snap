// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monster SpeedBonuses Small snapshot matches snapshot 1`] = `
<WithStyles(ExpansionPanel)
  CollapseProps={
    Object {
      "timeout": 100,
    }
  }
  defaultExpanded={true}
>
  <WithStyles(ExpansionPanelSummary)
    expandIcon={<pure(ExpandMoreIcon) />}
  >
    <WithStyles(Typography)
      className="heading"
    >
      Speed Bonuses
    </WithStyles(Typography)>
  </WithStyles(ExpansionPanelSummary)>
  <WithStyles(ExpansionPanelDetails)>
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <WithStyles(Grid)
        item={true}
        key="SpeedLand"
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="SpeedLand"
          >
            Speed Land
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="SpeedLand-helper-text"
            id="SpeedLand"
            name="SpeedLand"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="SpeedLand-helper-text"
          />
          <withHelpersModifiers(Help)>
            The distance in feet that the monster can walk in 1 round.
          </withHelpersModifiers(Help)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="SpeedSwim"
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="SpeedSwim"
          >
            Speed Swim
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="SpeedSwim-helper-text"
            id="SpeedSwim"
            name="SpeedSwim"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="SpeedSwim-helper-text"
          />
          <withHelpersModifiers(Help)>
            The distance in feet that the monster can swim in 1 round.
          </withHelpersModifiers(Help)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ExpansionPanelDetails)>
</WithStyles(ExpansionPanel)>
`;
